### ISSUE TITLE
Optimization of wd in separation case.

### ISSUE THREAD

**Issue by jonkoi** [holyseven/TransferLearningClassification#3]
Hi,

In the case where wd is always optimized with sgd in function _separate_reg_train_op. I can only see optimization of loss.

Am I missing something here?

**Comment by holyseven** [comment]
I did write a customized momentum optimizer, but not sure it is correct. So I removed from the code. It will be very nice if you can help. 

``` python
from tensorflow.python.ops import control_flow_ops
from tensorflow.python.ops import math_ops
from tensorflow.python.ops import state_ops
from tensorflow.python.framework import ops
from tensorflow.python.training import optimizer
import tensorflow as tf
import numpy as np


class CustomizedSGDMomentumOptimizer(optimizer.Optimizer):

    def __init__(self, learning_rate=0.01, wd_mode=4,
                 wd1=0.001, wd2=0.001, momentum=0.9,
                 new_layers_names=[], fine_tune_filename=None,
                 use_locking=False, name="CustomizedSGDMomentumOptimizer"):

        super(CustomizedSGDMomentumOptimizer, self).__init__(use_locking, name)
        self._lr = learning_rate
        self._wd1 = wd1
        self._wd2 = wd2
        self._momentum = momentum
        self._wdmode = wd_mode
        self.new_layers_names = new_layers_names
        self.fine_tune_filename = fine_tune_filename

    def _prepare(self):
        self._lr_t = ops.convert_to_tensor(self._lr, name="learning_rate")
        self._wd1_t = ops.convert_to_tensor(self._wd1, name="alpha_t")
        self._wd2_t = ops.convert_to_tensor(self._wd2, name="beta_t")
        self._momentum_t = ops.convert_to_tensor(self._momentum, name="momentum_t")
        self.reader = tf.train.NewCheckpointReader(self.fine_tune_filename)

    def _create_slots(self, var_list):
        for v in var_list:
            self._zeros_slot(v, "momentum", self._name)
        return

    def _apply_dense(self, grad, var):
        accum = self.get_slot(var, "momentum")

        lr_t = math_ops.cast(self._lr_t, var.dtype.base_dtype)
        wd1_t = math_ops.cast(self._wd1_t, var.dtype.base_dtype)
        wd2_t = math_ops.cast(self._wd2_t, var.dtype.base_dtype)
        momentum_t = math_ops.cast(self._momentum_t, var.dtype.base_dtype)

        # accumulate the momentum.
        accum = accum * momentum_t + grad

        # for layers whose parameters do not need to be regularized, just update with accumulated gradients.
        if 'weights' not in var.name:
            print 'no regularization:', var.name
            delta_var = lr_t * accum
            var_update = state_ops.assign_sub(var, delta_var)
            return control_flow_ops.group(*[var_update])

        # for new layers, l_2^2 regularization.
        if any(elem in var.name for elem in self.new_layers_names):  # L2 for new layers
            print 'apply l2:', var.name
            delta_var = lr_t * (accum + var * wd2_t)
            var_update = state_ops.assign_sub(var, delta_var)
            return control_flow_ops.group(*[var_update])

        # for existing weight layers, different regularization methods.
        name = var.name.split(':')[0]
        pre_trained_weights = self.reader.get_tensor(name)

        if self._wdmode == 0:
            print 'apply l2:', var.name
            grad_wd = var * wd1_t
        elif self._wdmode == 1:
            print 'apply l2-sp:', var.name
            grad_wd = (var - pre_trained_weights) * wd1_t
        elif self._wdmode == 4:
            print 'apply l1-sp:', var.name
            grad_wd = tf.sign(var-pre_trained_weights) * wd1_t
        elif self._wdmode == 5:
            print 'apply gl-sp:', var.name
            raise NotImplementedError("Group-Lasso is not implemented yet.")
        else:
            raise NotImplementedError("Unknown wd_mode, or not implemented yet.")

        delta_var = lr_t * (accum + grad_wd)
        var_update = state_ops.assign_sub(var, delta_var)
        return control_flow_ops.group(*[var_update])

    def _apply_sparse(self, grad, var):
        raise NotImplementedError("Sparse gradient updates are not supported.")
```
